core/src/main/java/io/questdb/cairo/O3PartitionJob.java

```diff
@@ -988,27 +1029,65 @@ private static void publishOpenColumnTasks(
 
         final long timestampMergeIndexAddr;
         final long timestampMergeIndexSize;
+        final TableRecordMetadata metadata = tableWriter.getMetadata();
         if (mergeType == O3_BLOCK_MERGE) {
-            timestampMergeIndexSize = (mergeDataHi - mergeDataLo + 1) * TIMESTAMP_MERGE_ENTRY_BYTES;
-            assert timestampMergeIndexSize > 0; // avoid SIGSEGV
+            long tempIndexSize = (mergeOOOHi - mergeOOOLo + 1 + mergeDataHi - mergeDataLo + 1) * TIMESTAMP_MERGE_ENTRY_BYTES;
+            assert tempIndexSize > 0; // avoid SIGSEGV
 
-            timestampMergeIndexAddr = createMergeIndex(
-                    srcTimestampAddr,
-                    sortedTimestampsAddr,
-                    mergeDataLo,
-                    mergeDataHi,
-                    mergeOOOLo,
-                    mergeOOOHi,
-                    timestampMergeIndexSize
-            );
+            if (!tableWriter.isDeduplicationEnabled()) {
+                timestampMergeIndexSize = tempIndexSize;
+                assert timestampMergeIndexSize > 0; // avoid SIGSEGV
+                timestampMergeIndexAddr = createMergeIndex(
+                        srcTimestampAddr,
+                        sortedTimestampsAddr,
+                        mergeDataLo,
+                        mergeDataHi,
+                        mergeOOOLo,
+                        mergeOOOHi,
+                        timestampMergeIndexSize
+                );
+            } else {
+                final long tempIndexAddr = Unsafe.malloc(tempIndexSize, MemoryTag.NATIVE_O3);
+                try {
+                    final long dedupRows = getDedupRows(
+                            srcTimestampAddr,
+                            mergeDataLo,
+                            mergeDataHi,
+                            sortedTimestampsAddr,
+                            mergeOOOLo,
+                            mergeOOOHi,
+                            tempIndexAddr
+                    );
+                    timestampMergeIndexSize = dedupRows * TIMESTAMP_MERGE_ENTRY_BYTES;
+                    timestampMergeIndexAddr = Unsafe.realloc(tempIndexAddr, tempIndexSize, timestampMergeIndexSize, MemoryTag.NATIVE_O3);
+                    final long duplicateCount = (mergeOOOHi - mergeOOOLo + 1 + mergeDataHi - mergeDataLo + 1) - dedupRows;
+                    newPartitionSize -= duplicateCount;
+                    if (oldPartitionTimestamp == partitionTimestamp) {
+                        oldPartitionSize -= duplicateCount;
+                    }
+                } catch (Throwable e) {
+                    tableWriter.o3BumpErrorCount();
+                    LOG.error().$("open column error [table=").utf8(tableWriter.getTableToken().getTableName())
```

Comment:
should this be critical?

Reply:
fixed in [f339126](https://github.com/questdb/questdb/pull/3524/commits/f339126ea9719c6207530d6e8a49942a502e23c7)